cardano-crypto-class/src/Cardano/Crypto/SECP256K1/C.hs

```diff
@@ -12,18 +13,51 @@ module Cardano.Crypto.SECP256K1.C (
   secpSchnorrSigSignCustom,
   secpKeyPairXOnlyPub,
   secpSchnorrSigVerify,
+  secpXOnlyPubkeySerialize,
+  secpXOnlyPubkeyParse,
+  secpCtxPtr,
+  secpEcPubkeyCreate,
+  secpEcdsaSign,
+  secpEcdsaVerify,
+  secpEcCompressed,
+  secpEcPubkeySerialize,
+  secpEcdsaSignatureSerializeCompact,
+  secpEcdsaSignatureParseCompact,
+  secpEcPubkeyParse,
   ) where
 
 import Data.Bits ((.|.))
 import Foreign.Ptr (Ptr)
-import Foreign.C.Types (CUChar, CSize (CSize), CInt (CInt))
-import Cardano.Foreign
-import Cardano.Crypto.SECP256K1.Constants
+import System.IO.Unsafe (unsafePerformIO)
+import Data.Word (Word8)
+import Foreign.C.Types (CUChar, CSize (CSize), CInt (CInt), CChar)
+import Cardano.Foreign (SizedPtr (SizedPtr))
+import Cardano.Crypto.SECP256K1.Constants (
+  SECP256K1_SCHNORR_KEYPAIR_BYTES,
+  SECP256K1_SCHNORR_PRIVKEY_BYTES,
+  SECP256K1_SCHNORR_SIGNATURE_BYTES,
+  SECP256K1_SCHNORR_PUBKEY_BYTES_INTERNAL,
+  SECP256K1_SCHNORR_PUBKEY_BYTES,
+  SECP256K1_ECDSA_PUBKEY_BYTES_INTERNAL,
+  SECP256K1_ECDSA_PRIVKEY_BYTES,
+  SECP256K1_ECDSA_SIGNATURE_BYTES_INTERNAL,
+  SECP256K1_ECDSA_SIGNATURE_BYTES,
+  )
 
 data SECP256k1Context
 
 data SECP256k1SchnorrExtraParams
 
+-- We create a single context for both signing and verifying, which we use
+-- everywhere. This saves considerable time, and is safe, provided nobody
+-- outside cardano-base gets to touch it.
+--
+-- We do _not_ make this dupable, as the whole point is _not_ to compute it more
+-- than once!
+{-# NOINLINE secpCtxPtr #-}
+secpCtxPtr :: Ptr SECP256k1Context
+secpCtxPtr = unsafePerformIO . secpContextCreate $ secpContextSignVerify
```

Comment:
I agree with all the above, however, if we expose such a function, we'd _definitely_ need @michaelpj to agree, since this would significantly change how the functionality here gets used. This is especially because nothing else in `DSIGN` needs anything of this nature.

How do we address `libsodium` initialization for the stuff we have that uses it? I think it's a similar problem.

Reply:
This might also affect the costing of the function in plutus. However, I guess that we setup a node once, and run the function several times to then get the average and use it as the function cost. Is that the case?

Reply:
The C library says the following: 

```
Do not create a new context object for each operation, as construction is
far slower than all other API calls (~100 times slower than an ECDSA
verification).
```

Reply:
In an ideal world, I would recommend having an `IO` operation to do this initialization, and require every application's `main` to call it.

Judging by @iquerejeta's response, rebuilding the context on every call would be unacceptably slow.

Reply:
@kozross I also agree that ideally the context would be part of the environment that is threaded through the application. However, I assume it will be a bit invasive, although maybe @michaelpj can correct me on this. Most importantly we can  get away with only one instance of the context for the whole application, since it is never mutated and thus is thread safe, I think it is acceptable to use this approach, but we do need an IO action that forces that pointer, so we do not run into this scenario where one secp function call is extremely expensive when compared to others. I am sure we can find a place upon a node start up where we can such initializing action.

Reply:
'Fully threading around' the context is a bit unviable here: we only support passing contextual information for signing and verifying, but we would need this literally everywhere. This would require too large a redesign. My suggestion is halfway: have a function like `setUpCardanoCrypto :: IO ()` which does all the setup, including building and forcing our context (and initializing `libsodium` while we're at it). This won't require an API change, but _will_ require extra work in any application using `cardano-crypto-class`.

Reply:
I don't have anything to oppose to your suggestion, I just don't see what initialisation libsodium requires.

Reply:
@iquerejeta - someone has to call `sodium_init` somewhere.

Reply:
ah, yes, indeed libsodium also requires initialisation with that function 👍  you can forget my comment then. Just for reference, [here](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/Libsodium/Init.hs) is the init function.

Reply:
We still haven't addressed the fact that the context is never cleaned up. It is more than likely totally fine for cardano-node, since we want this context to live as long as the node is alive, but morally it is wrong since it is a memory leak by definition, because this context is never GCed/deallocated. Using a `ForeignPtr` with a finalizer solves that.

Reply:
Same bug for `sodiumInit`?

Reply:
Stupid question: if we have a global pointer to the context, will it ever actually get GC'd? So will the finalizer actually run?

Reply:
> if we have a global pointer to the context, will it ever actually get GC'd? So will the finalizer actually run?

@michaelpj This is a very good question. Answer is it will be GCed as soon as there are no references to it. However, I wanted to convince myself that this is indeed true, so here is an example:
```haskell
import Control.Concurrent
import Control.Exception
import Foreign.Marshal.Alloc
import Foreign.Ptr
import Foreign.Storable
import GHC.Weak
import System.IO.Unsafe
import System.Mem

topLevelPtr :: Ptr Int
topLevelPtr = unsafePerformIO $ do
  ptr <- malloc
  poke ptr 2022
  putStrLn "Allocated the Ptr"
  _ <- mkWeak ptr ptr $ Just $ free ptr >> putStrLn "Finalizer executed"
  pure ptr
{-# NOINLINE topLevelPtr #-}

main :: IO ()
main = do
  putStrLn "Started the App. Now will force the top level pointer"
  _ <- evaluate topLevelPtr
  putStrLn "Forced the top level pointer. Now will read from it"
  v <- peek topLevelPtr
  putStrLn $ "Read the value from the pointer: " ++ show v ++ ". Now will check if GC cleans it up"
  performGC
  putStrLn "Waiting for 1 sec"
  threadDelay 1000000
  putStrLn "Waiting for another sec"
  threadDelay 1000000
  putStrLn "Done"
```

If we run it we will get finalizer executing before the final line of code is performed, thus before the application is done:
```
Started the App. Now will force the top level pointer
Allocated the Ptr
Forced the top level pointer. Now will read from it
Read the value from the pointer: 2022. Now will check if GC cleans it up
Waiting for 1 sec
Finalizer executed
Waiting for another sec
Done
```

Reply:
@lehins So does this mean we want an explicit finalizer in a `ForeignPtr` in our case or not?

Reply:
> So does this mean we want an explicit finalizer in a ForeignPtr in our case or not?

@kozross Yes we do, that's precisely what I said in my original [comment](https://github.com/input-output-hk/cardano-base/pull/289#discussion_r923885578), except we should use the ForeignPtr finalizer mechanism, rather than the Weak pointer from my example above:

```haskell
secpCtxPtr :: ForeignPtr SECP256k1Context
secpCtxPtr = unsafePerformIO $ mask_ $ do
 ctx <- secpContextCreate secpContextSignVerify
 newForeignPtr secpContextDestroy ctx

foreign import unsafe ccall "secp256k1.h &secp256k1_context_destroy"
  secpContextDestroy :: FinalizerPtr SECP256k1Context
```

Reply:
OK, can do.